File: app/application/repositories/analysis_job.py
"""Repository for AnalysisJob persistence.

Provides CRUD operations and queue-management queries for the analysis
job table used by the ML pipeline scheduler.
"""

from datetime import datetime

from sqlalchemy import delete, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.models import AnalysisJob


class AnalysisJobRepository:
    """Provide CRUD and query access to AnalysisJob records.

    Args:
        session: The async SQLAlchemy session to use for all operations.
    """

    def __init__(self, session: AsyncSession):
        self._session = session

    async def delete_by_snapshot_ids(self, snapshot_ids: list[int]) -> int:
        """Delete all analysis jobs for the given snapshot IDs.

        Args:
            snapshot_ids: Snapshot primary keys whose jobs should be removed.

        Returns:
            The number of analysis jobs deleted.
        """
        if not snapshot_ids:
            return 0
        result = await self._session.execute(
            delete(AnalysisJob).where(AnalysisJob.snapshot_id.in_(snapshot_ids))
        )
        await self._session.flush()
        return result.rowcount or 0

    async def add(self, job: AnalysisJob) -> None:
        """Persist a new analysis job to the database.

        Args:
            job: The unsaved AnalysisJob instance.
        """
        self._session.add(job)

    async def get_by_id(self, job_id: int) -> AnalysisJob | None:
        """Fetch a single job by its primary key.

        Args:
            job_id: The unique identifier of the job.

        Returns:
            The AnalysisJob if found, otherwise None.
        """
        result = await self._session.execute(
            select(AnalysisJob).where(AnalysisJob.id == job_id)
        )
        return result.scalar_one_or_none()

    async def get_pending(self, limit: int = 10) -> list[AnalysisJob]:
        """Fetch the next batch of pending jobs ordered by priority and age.

        Args:
            limit: Maximum number of jobs to return.

        Returns:
            List of pending AnalysisJob records.
        """
        result = await self._session.execute(
            select(AnalysisJob)
            .where(AnalysisJob.status == "pending")
            .order_by(AnalysisJob.priority.desc(), AnalysisJob.requested_at)
            .limit(limit)
        )
        return list(result.scalars().all())

    async def mark_started(self, job_id: int) -> None:
        """Transition a job from pending to processing and increment attempts.

        Args:
            job_id: The job to mark as started.
        """
        await self._session.execute(
            update(AnalysisJob)
            .where(AnalysisJob.id == job_id)
            .values(status="processing", started_at=datetime.now(), attempts=AnalysisJob.attempts + 1)
        )

    async def mark_completed(self, job_id: int) -> None:
        """Mark a processing job as completed with a finished timestamp.

        Args:
            job_id: The job to mark as completed.
        """
        await self._session.execute(
            update(AnalysisJob)
            .where(AnalysisJob.id == job_id)
            .values(status="completed", finished_at=datetime.now())
        )

    async def mark_failed(self, job_id: int, error_message: str) -> None:
        """Mark a processing job as failed with an error description.

        Args:
            job_id: The job to mark as failed.
            error_message: Human-readable description of what went wrong.
        """
        await self._session.execute(
            update(AnalysisJob)
            .where(AnalysisJob.id == job_id)
            .values(status="failed", error_message=error_message, finished_at=datetime.now())
        )

File: internal/repository/analysis_job.go
package repository

import (
	"context"
	"time"

	"github.com/jmoiron/sqlx"

	"github.com/Lmex89/home-cameras/internal/domain"
)

// AnalysisJobRepository persists analysis jobs and queue state.
type AnalysisJobRepository struct {
	db DBTX
}

// NewAnalysisJobRepository builds an analysis job repository.
func NewAnalysisJobRepository(db DBTX) *AnalysisJobRepository { return &AnalysisJobRepository{db: db} }

const jobCols = `id, snapshot_id, job_type, status, priority, attempts, max_attempts,
	error_message, requested_at, started_at, finished_at`

// Add inserts a new pending job and populates its id.
func (r *AnalysisJobRepository) Add(ctx context.Context, j *domain.AnalysisJob) error {
	res, err := r.db.ExecContext(ctx,
		`INSERT INTO analysis_jobs (snapshot_id, job_type, status, priority)
		 VALUES (?, ?, ?, ?)`,
		j.SnapshotID, j.JobType, j.Status, j.Priority)
	if err != nil {
		return err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return err
	}
	j.ID = id
	return nil
}

// GetByID returns one job.
func (r *AnalysisJobRepository) GetByID(ctx context.Context, id int64) (*domain.AnalysisJob, error) {
	var j domain.AnalysisJob
	err := sqlx.GetContext(ctx, r.db, &j,
		"SELECT "+jobCols+" FROM analysis_jobs WHERE id = ?", id)
	if err != nil {
		return nil, err
	}
	return &j, nil
}

// GetPending returns pending jobs by priority then age.
func (r *AnalysisJobRepository) GetPending(ctx context.Context, limit int) ([]domain.AnalysisJob, error) {
	var jobs []domain.AnalysisJob
	err := sqlx.SelectContext(ctx, r.db, &jobs,
		"SELECT "+jobCols+` FROM analysis_jobs
		 WHERE status = 'pending'
		 ORDER BY priority DESC, requested_at LIMIT ?`, limit)
	return jobs, err
}

// MarkStarted transitions a job to processing and bumps attempts.
func (r *AnalysisJobRepository) MarkStarted(ctx context.Context, id int64) error {
	now := time.Now().Format("2006-01-02 15:04:05")
	_, err := r.db.ExecContext(ctx,
		`UPDATE analysis_jobs SET status = 'processing', started_at = ?,
		 attempts = attempts + 1 WHERE id = ?`, now, id)
	return err
}

// MarkCompleted marks a job done.
func (r *AnalysisJobRepository) MarkCompleted(ctx context.Context, id int64) error {
	now := time.Now().Format("2006-01-02 15:04:05")
	_, err := r.db.ExecContext(ctx,
		"UPDATE analysis_jobs SET status = 'completed', finished_at = ? WHERE id = ?", now, id)
	return err
}

// MarkFailed marks a job failed with a message.
func (r *AnalysisJobRepository) MarkFailed(ctx context.Context, id int64, msg string) error {
	now := time.Now().Format("2006-01-02 15:04:05")
	_, err := r.db.ExecContext(ctx,
		"UPDATE analysis_jobs SET status = 'failed', error_message = ?, finished_at = ? WHERE id = ?",
		msg, now, id)
	return err
}

// DeleteBySnapshotIDs removes jobs for the given snapshots.
func (r *AnalysisJobRepository) DeleteBySnapshotIDs(ctx context.Context, ids []int64) (int64, error) {
	if len(ids) == 0 {
		return 0, nil
	}
	query, args, err := sqlx.In("DELETE FROM analysis_jobs WHERE snapshot_id IN (?)", ids)
	if err != nil {
		return 0, err
	}
	res, err := r.db.ExecContext(ctx, r.db.Rebind(query), args...)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}
